src/hooks/use-deck-editor.ts

```diff
@@ -0,0 +1,193 @@
+    addedCards: {},
+    updatedCards: {},
+    deletedCards: {},
+  });
+
+  function save() {
+    dispatch({ type: DECK_REDUCER_TYPE.SAVE });
+  }
+  function discardChanges() {
+    dispatch({ type: DECK_REDUCER_TYPE.DISCARD_CHANGES });
+  }
+  function updateMetaData(metaData: DeckMetaData) {
+    dispatch({ type: DECK_REDUCER_TYPE.UPDATE_META_DATA, metaData });
+  }
+  function addCard() {
+    dispatch({ type: DECK_REDUCER_TYPE.ADD_CARD });
+  }
+  function deleteCard(card: Card) {
+    dispatch({ type: DECK_REDUCER_TYPE.DELETE_CARD, card });
+  }
+  function updateCard(card: Card) {
+    dispatch({ type: DECK_REDUCER_TYPE.UPDATE_CARD, card });
+  }
+  function reorderCards(cards: Card[]) {
+    dispatch({ type: DECK_REDUCER_TYPE.REORDER_CARDS, cards });
+  }
+
+  return [
+    { deck: state.currentDeck, hasUnsavedChanges: state.hasUnsavedChanges },
+    { addCard, deleteCard, updateCard, updateMetaData, save, discardChanges, reorderCards },
+  ];
+};
+
+function deckEditorReducer(state: DeckReducerState, action: DeckReducerDispatch): DeckReducerState {
+  const { currentDeck, savedDeck, addedCards, deletedCards, updatedCards } = state;
+  const { type, card, cards, metaData } = action;
+
+  switch (type) {
+    case DECK_REDUCER_TYPE.ADD_CARD:
+      const newCard = createNewCard();
+      return {
+        ...state,
+        currentDeck: addCardToDeck(newCard, currentDeck),
+        addedCards: addCardToMap(newCard, addedCards),
+        hasUnsavedChanges: true,
+      };
+
+    case DECK_REDUCER_TYPE.DELETE_CARD:
+      if (card === undefined) {
+        throw new Error('action.card must not be undefined');
+      }
+      return {
+        ...state,
+        addedCards: removeCardFromMap(card, addedCards),
+        updatedCards: removeCardFromMap(card, updatedCards),
+        deletedCards: addCardToMap(card, deletedCards),
+        currentDeck: removeCardFromDeck(card, currentDeck),
+        hasUnsavedChanges: true,
+      };
+
+    case DECK_REDUCER_TYPE.SAVE:
+      // Todo: make the API calls
+      // Todo: update the id's of new cards
+      return {
+        ...state,
+        savedDeck: currentDeck,
+        addedCards: {},
+        deletedCards: {},
+        updatedCards: {},
+        hasUnsavedChanges: false,
+      };
+
+    case DECK_REDUCER_TYPE.DISCARD_CHANGES:
+      return {
+        ...state,
+        currentDeck: savedDeck,
+        addedCards: {},
+        deletedCards: {},
+        updatedCards: {},
+        hasUnsavedChanges: false,
+      };
+
+    case DECK_REDUCER_TYPE.UPDATE_META_DATA:
+      if (metaData === undefined) throw new Error('action.metaData must not be undefined');
+
+      return { ...state, currentDeck: { ...currentDeck, metaData }, hasUnsavedChanges: true };
+
+    case DECK_REDUCER_TYPE.UPDATE_CARD:
+      if (card === undefined) throw new Error('action.card must not be undefined');
+
+      return {
+        ...state,
+        addedCards: card.id ? addCardToMap(card, addedCards) : addedCards,
```

Comment:
i believe the logic is reversed here? 

if a card was updated and has an id, that means the card came from the backend; so we shouldn't add it to the CardsMap.

Reply:
Thanks for the callout. I made a last-minute change to check if new and get the logic flipped 😅. Fixed now.